### PR TITLE
Improve sync performance with no-op when possible

### PR DESCRIPTION
**Pull request by fao89**
https://pulp.plan.io/issues/7942
closes #7942

### CODE REVIEW

File: pulp_ansible/app/viewsets.py [hunk 1]
```diff
@@ -245,6 +245,11 @@ class CollectionRemoteViewSet(RemoteViewSet):
     queryset = CollectionRemote.objects.all()
     serializer_class = CollectionRemoteSerializer
 
+    def update(self, request, pk, **kwargs):
+        if "requirements_file" in request.data:
+            request.data["last_synced_metadata_time"] = None
```
Comment (fao89) [inline pulp_ansible/app/viewsets.py:250]: if we change the `requirements_file` unset `last_synced_metadata_time`.

for avoiding no-op when a new requirements file contains more collections than the previous one

File: pulp_ansible/app/tasks/collections.py [hunk 2]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+                pass
+
+            collection_info = self.collection_info or []
+            sources = {r.source for r in collection_info if r.source}
+            sources.add(self.remote.url)
+
+            published = metadata.get("published")
+            if published == self.remote.last_synced_metadata_time and len(sources) < 2:
+                log.info(_("no-op"))
+                return
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:532]: we may have `self.remote.url` and `sources` pointing to different repos, should we only consider `self.remote.url` for no-op?

**Comment by bmbouter** [comment]
@fao89 maybe we should break this story into two issues: one for just using the sync endpoint and then approach no-op separately? I'm ok to keep it together also. Wdyt?

**Comment by fao89** [comment]
> @fao89 maybe we should break this story into two issues: one for just using the sync endpoint and then approach no-op separately? I'm ok to keep it together also. Wdyt?

I can rename this PR and it would be the no-op PR

File: pulp_ansible/app/viewsets.py (hunk 1 above)
Comment (bmbouter) [reply pulp_ansible/app/viewsets.py:251]: I think this is good. I didn't think about this. :+1:

Comment (bmbouter) [reply pulp_ansible/app/viewsets.py:251]: Can a test be added for this part please?

File: pulp_ansible/app/tasks/collections.py [hunk 3]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+
+            self.last_synced_metadata_time = published
+
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:535]: I think this should be moved to its own function that is called prior to `_fetch_collection_version_metadata`. I think we want to keep the function lengths small (and responsibilities separated) to keep things manageable.

File: pulp_ansible/app/tasks/collections.py [hunk 4]
```diff
@@ -304,12 +308,13 @@ def __init__(self, remote):
...
+        self.last_synced_metadata_time = None
 
         # Interpret download policy
         self.deferred_download = self.remote.policy != Remote.IMMEDIATE
 
     @alru_cache(maxsize=128)
-    async def _get_collection_api(self, root):
+    async def _get_collection_api_root(self, root):
         """
         Returns the collection api path and api version.
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:319]: This docstring is out of date now I think.

File: pulp_ansible/app/tasks/collections.py [hunk 5]
```diff
@@ -340,10 +345,14 @@ async def _get_collection_api(self, root):
         else:
             raise RuntimeError(_("Unsupported API versions at {}").format(root))
 
-        endpoint = f"{root}v{api_version}/collections/"
+        endpoint = f"{root}v{api_version}"
 
         return endpoint, api_version
 
+    async def _get_collection_api(self, root):
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:352]: This needs a docstring I think.

File: pulp_ansible/app/tasks/collections.py [hunk 6]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+        if api_version == 3:
+            downloader = self.remote.get_downloader(
+                url=root, silence_errors_for_response_status_codes={404}
+            )
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                pass
+
+            collection_info = self.collection_info or []
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:525]: I don't expect an `or []` to be needed here. Is it needed?

File: pulp_ansible/app/tasks/collections.py [hunk 7]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+            sources = {r.source for r in collection_info if r.source}
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:526]: Can this be unrolled to be more readable. I had to read it 3 times and while it's allowed I think we want code to be super readable.

File: pulp_ansible/app/tasks/collections.py [hunk 8]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+            sources.add(self.remote.url)
+
+            published = metadata.get("published")
+            if published == self.remote.last_synced_metadata_time and len(sources) < 2:
+                log.info(_("no-op"))
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:531]: Can this log message be expanded some. It would be good if it also logged if it's being no-opped due to too many sources, versus the published time not changing.

File: pulp_ansible/app/tasks/collections.py (hunk 2 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:532]: Yeah maybe we should leave off the inspection of sources from this PR at least to keep it simple from the start... wdyt?

File: pulp_ansible/app/tasks/collections.py [hunk 9]
```diff
@@ -503,6 +512,27 @@ async def run(self):
...
+                return
+
+            self.last_synced_metadata_time = published
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:534]: Does this need to be saved?

**Comment by bmbouter** [comment]
I think another part should be added to this PR, a new, optional sync parameter called `force`. When `force=True` the no-op is skipped no matter what. When `force=False` a no-op could occur. If unspecified the sync defaults to `force=False`.

I think having the `force` option as a sync param will allow us to not worry so much about the `sources` corner cases. If a user is getting a no-op and they don't want one, they just call with `force=True`.

File: pulp_ansible/app/tasks/collections.py (hunk 9 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:534]: for successful sync, we want to keep the published time to compare with future syncs

Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:534]: I agree, I'm asking where is the call to `self.save()`

File: pulp_ansible/app/tasks/collections.py [hunk 10]
```diff
@@ -84,6 +84,10 @@ def sync(remote_pk, repository_pk, mirror):
     d_version = AnsibleDeclarativeVersion(first_stage, repository, mirror=mirror)
     d_version.create()
 
+    if first_stage.last_synced_metadata_time:
+        remote.last_synced_metadata_time = first_stage.last_synced_metadata_time
+        remote.save()
```
Comment (fao89) [inline pulp_ansible/app/tasks/collections.py:89]: @bmbouter I only save published time when the sync is successful

**Comment by fao89** [comment]
> I think another part should be added to this PR, a new, optional sync parameter called `force`. When `force=True` the no-op is skipped no matter what. When `force=False` a no-op could occur. If unspecified the sync defaults to `force=False`.
> 
> I think having the `force` option as a sync param will allow us to not worry so much about the `sources` corner cases. If a user is getting a no-op and they don't want one, they just call with `force=True`.

on pulp_rpm we specify `optimize=True`, I changed to this approach

File: pulp_ansible/tests/functional/api/collection/v3/test_sync.py [hunk 11]
```diff
@@ -24,32 +24,54 @@ class SyncCollectionsFromPulpServerTestCase(TestCaseUsingBindings, SyncHelpersMi
...
+            requirements_file=self.requirements_file,
+        )
+        second_remote = self.remote_collection_api.create(second_body)
+        self.addCleanup(self.remote_collection_api.delete, second_remote.pulp_href)
+
+        second_repo = self._create_repo_and_sync_with_remote(second_remote)
+
+        # Resync
+        self._sync_repo(second_repo, remote=second_remote.pulp_href, optimize=True)
+
```
Comment (fao89) [inline pulp_ansible/tests/functional/api/collection/v3/test_sync.py:71]: I don't know how to assert the log, the test run on client side and the log happens on server side

File: pulp_ansible/app/models.py [hunk 12]
```diff
@@ -235,6 +235,8 @@ class AnsibleRepository(Repository):
     CONTENT_TYPES = [Role, CollectionVersion]
     REMOTE_TYPES = [RoleRemote, CollectionRemote]
 
+    last_synced_metadata_time = models.DateTimeField(null=True)
```
Comment (bmbouter) [inline pulp_ansible/app/models.py:238]: I think we should add this to the docstring

File: pulp_ansible/app/viewsets.py [hunk 13]
```diff
@@ -245,6 +246,19 @@ class CollectionRemoteViewSet(RemoteViewSet):
...
+        if "url" in request.data or "requirements_file" in request.data:
+            repos = list(
+                AnsibleRepository.objects.filter(
+                    remote_id=pk, last_synced_metadata_time__isnull=False
+                ).all()
+            )
+            for repo in repos:
+                repo.last_synced_metadata_time = None
+            AnsibleRepository.objects.bulk_update(repos, ["last_synced_metadata_time"])
+        return super().update(request, pk, **kwargs)
```
Comment (bmbouter) [inline pulp_ansible/app/viewsets.py:260]: I think this should all be wrapped in one db transaction.

File: CHANGES/7942.feature [hunk 14]
```diff
@@ -0,0 +1 @@
+Improve sync performance with no-op when possible.
```
Comment (bmbouter) [inline CHANGES/7942.feature:1]: Please add the fact that the no-op can be disabled with the `optimize=False` sync parameter also.

File: pulp_ansible/app/tasks/collections.py [hunk 15]
```diff
@@ -608,13 +626,60 @@ async def _find_all_collections(self):
...
+        return False
+
     async def run(self):
         """
         Build and emit `DeclarativeContent` from the ansible metadata.
         """
+        if self.repository.remote and self.remote.pk == self.repository.remote.pk:
+            can_be_optimized = await self._can_be_optimized()
+            if self.optimize and can_be_optimized:
+                return
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:676]: It would be good to add a logging statement at debug level here indicating that the no-op occurred also.

File: pulp_ansible/app/tasks/collections.py (hunk 9 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:665]: So where does this self object get saved?

File: CHANGES/7942.feature (hunk 14 above)
Comment (fao89) [reply CHANGES/7942.feature:1]: it is disabled by default, we need to set it to true to enable

Comment (bmbouter) [reply CHANGES/7942.feature:1]: Oh I think optimize should be enabled by default.

Regardless though, we need to identify that this new option is available to the user.

File: pulp_ansible/app/viewsets.py (hunk 13 above)
Comment (fao89) [reply pulp_ansible/app/viewsets.py:260]: super() triggers a task, not sure if it would be effective to add it

File: CHANGES/7942.feature (hunk 14 above)
Comment (bmbouter) [reply CHANGES/7942.feature:1]: I think we definitly want it enabled by default. This is also how other content types work, e.g. rpm.

File: pulp_ansible/tests/functional/api/collection/v3/test_sync.py [hunk 16]
```diff
@@ -75,6 +86,53 @@ def test_sync_collections_from_pulp_using_mirror_second_time(self):
         second_content = self.cv_api.list(repository_version=f"{second_repo.pulp_href}versions/1/")
         self.assertGreaterEqual(len(second_content.results), 1)
 
+    def test_noop_resync_collections_from_pulp(self):
+        """Test sync collections from pulp server."""
```
Comment (bmbouter) [inline pulp_ansible/tests/functional/api/collection/v3/test_sync.py:90]: Can this be more descriptive. This seems not accurate for this test.

File: pulp_ansible/app/tasks/collections.py [hunk 17]
```diff
@@ -608,13 +626,61 @@ async def _find_all_collections(self):
...
+        return False
+
     async def run(self):
         """
         Build and emit `DeclarativeContent` from the ansible metadata.
         """
+        if self.repository.remote and self.remote.pk == self.repository.remote.pk:
+            can_be_optimized = await self._can_be_optimized()
+            if self.optimize and can_be_optimized:
+                log.debug("no-op: repmote wasn't updated since last sync.")
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:676]:
```suggestion
                log.debug(_("no-op: remote data has not changed since last sync."))
```

File: pulp_ansible/app/tasks/collections.py [hunk 18]
```diff
@@ -608,13 +626,61 @@ async def _find_all_collections(self):
...
+            )
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                return False
+
+            sources = set()
+            if self.collection_info:
+                sources = {r.source for r in self.collection_info if r.source}
+            sources.add(self.remote.url)
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:644]: I think having the > 2 check right here are returning False would be better than leaving the variable to be reused further down in a more complex check.

File: pulp_ansible/app/tasks/collections.py [hunk 19]
```diff
@@ -608,13 +626,61 @@ async def _find_all_collections(self):
...
+
+            published = metadata.get("published")
+            repo_metadata_time = self.repository.last_synced_metadata_time
+            repo_metadata_time = repo_metadata_time.isoformat() if repo_metadata_time else ""
+
+            # From serializers.DateTimeField.to_representation:
+            if repo_metadata_time.endswith("+00:00"):
+                repo_metadata_time = repo_metadata_time[:-6] + "Z"
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:652]: How can this part be made to have `published and `repo_metadata_time` both be DateTime objects instead of strings? I think we want to use DateTime for comparisons when those strings represent time.

File: pulp_ansible/app/tasks/collections.py [hunk 20]
```diff
@@ -608,13 +626,61 @@ async def _find_all_collections(self):
...
+
+            if published == repo_metadata_time and len(sources) < 2:
+                msg = _(
+                    "no-op: {repo} did not change since last sync - {published}".format(
+                        repo=self.remote.url, published=published
+                    )
+                )
+                noop = ProgressReport(message=msg, code="noop")
+                noop.state = TASK_STATES.COMPLETED
+                noop.save()
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:662]: It I think would be more reliable to report to the user that the no-op check occured anytime it did occur. To do that I would move this part higher up in the function before any `return` statements (maybe even at the beginning).

File: pulp_ansible/app/viewsets.py [hunk 21]
```diff
@@ -245,6 +247,20 @@ class CollectionRemoteViewSet(RemoteViewSet):
     queryset = CollectionRemote.objects.all()
     serializer_class = CollectionRemoteSerializer
 
+    def update(self, request, pk, **kwargs):
```
Comment (bmbouter) [inline pulp_ansible/app/viewsets.py:250]: I forgot that this `super()` call runs inside a task. Since that's the case we need to do this differently for correctness reasons. I recommend doing these things:

1) Create a new task called `update_collection_remote` and have it contain this update code.
2) Instead of calling super() which would dispatch another task have your `update_collection_remote` task call the `tasks.base.general_update` [in the saw way that the super does](https://github.com/pulp/pulpcore/blob/5a6d5c5f2ebe06922319235b3c845608d10e6f3a/pulpcore/app/viewsets/base.py#L454).
3) You'll need to ensure the `update_collection_remote` task not only locks on the remote being updated but also the AnsibleRepository (this is the main correctness change also)
4) Since you're not calling super have your code return the `OperationPostponed`.
5) Also have you `update_collection_remote` use a Transaction so all postgresql operations happen or none of them do.

File: pulp_ansible/app/tasks/collections.py (hunk 19 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:652]: this is actually what this code is doing, converting the string it gets from serializer to datetime

File: pulp_ansible/app/tasks/collections.py (hunk 20 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:662]: this is what this code is doing, no-op only happens when return True, this is the only part when it returns True

Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:662]: What I'm saying is the progress reporting should say that it is checking for a no-op regardless of it if occurs or not. In terms of reporting on what it's doing I expect this progress report to be there when optimize=True and not to be there when optimize=False.

File: pulp_ansible/app/viewsets.py (hunk 21 above)
Comment (fao89) [reply pulp_ansible/app/viewsets.py:253]: it may be zero or more than one repository, so I don't think it is possible to lock on repository

File: pulp_ansible/app/tasks/collections.py (hunk 9 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:697]: https://github.com/pulp/pulp_ansible/pull/511#discussion_r566897461

File: pulp_ansible/app/tasks/collections.py (hunk 20 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:694]: so how can I assert if no-op happened?
I added the ProgressReport just for it

Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:694]: I expected it's not on asserting that the progress reporting shows no-op was checked or not, it's that the rest of the 'syncing, etc, etc' progress reports are not present and the state is `completed`.

File: pulp_ansible/app/tasks/collections.py [hunk 22]
```diff
@@ -294,23 +342,30 @@ class CollectionSyncFirstStage(Stage):
     The first stage of a pulp_ansible sync pipeline.
     """
 
-    def __init__(self, remote):
+    def __init__(self, remote, repository, optimize=False):
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:345]: I see the release note chnaged, should this also be `optimize=True` for the default?

File: pulp_ansible/app/tasks/collections.py [hunk 23]
```diff
@@ -608,13 +663,56 @@ async def _find_all_collections(self):
...
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                return False
+
+            sources = set()
+            if self.collection_info:
+                sources = {r.source for r in self.collection_info if r.source}
+            sources.add(self.remote.url)
+            has_multiple_sources = len(sources) > 1
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:682]: Can this just `return False` here?

File: pulp_ansible/app/tasks/collections.py [hunk 24]
```diff
@@ -608,13 +663,56 @@ async def _find_all_collections(self):
...
+
+            published = parse_datetime(metadata.get("published", ""))
+
+            if published == self.repository.last_synced_metadata_time and not has_multiple_sources:
+                msg = _(
+                    "no-op: {repo} did not change since last sync - {published}".format(
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:688]: I'm suggesting to make this a report that the no-op was checked, not that it occurred. So that would have it be reporting in all cases unless optimize=False at which point I expect Pulp not to evaluate the no-op case at all.

File: pulp_ansible/app/tasks/collections.py (hunk 22 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:345]: it gets True from here: https://github.com/fao89/pulp_ansible/blob/7942/pulp_ansible/app/serializers.py#L101
like pulp_rpm does

File: pulp_ansible/app/viewsets.py (hunk 21 above)
Comment (bmbouter) [reply pulp_ansible/app/viewsets.py:253]: Why couldn't it just lock on 0 or more also? Pulp supports multi-locking.

File: pulp_ansible/app/tasks/collections.py (hunk 22 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:345]: Got it! Then I guess I'm confused why it's a kwarg at all if there will always be a value specified from the serializer. Do you know why? (I don't).

File: CHANGES/7942.feature [hunk 25]
```diff
@@ -0,0 +1,2 @@
...
+For disabling no-op optimization set ``optimize=False``.
```
Comment (bmbouter) [inline CHANGES/7942.feature:2]: Can this be combine moved to a second sentence in the previous line and also:

```suggestion
To disabled the no-op optimization use the ``optimize=False`` option on the ``sync`` call.
```

File: pulp_ansible/app/tasks/collections.py (hunk 24 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:688]: I don't see the value of having no-op check in progress reports

File: pulp_ansible/app/tasks/collections.py (hunk 22 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:345]: Oh actually I see it's always specified in `CollectionSyncFirstStage(remote, repository, optimize)` so this should be an args.

File: pulp_ansible/app/tasks/collections.py (hunk 24 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:689]: If you want to take it out I'm ok with that. My thought was that if we don't have any progress report and the no-op runs the user has no info about why. Also it tells the user what was performed differently when they call `optimize=False`.

What we're doing right now though doesn't seem right (to me). Usually progress reports report on what occurred in the task code, but this is more like a message to the user about why the task reached its completed state.

File: pulp_ansible/app/viewsets.py [hunk 26]
```diff
@@ -200,30 +198,33 @@ class AnsibleRepositoryViewSet(RepositoryViewSet, ModifyRepositoryActionMixin):
...
-        serializer = RepositorySyncURLSerializer(
+        serializer = AnsibleRepositorySyncURLSerializer(
             data=request.data, context={"request": request, "repository_pk": repository.pk}
         )
         serializer.is_valid(raise_exception=True)
 
         remote = serializer.validated_data.get("remote", repository.remote)
         remote = remote.cast()
 
+        mirror = serializer.validated_data.get("mirror", False)
```
Comment (bmbouter) [inline pulp_ansible/app/viewsets.py:215]: This is similar, I don't think we should have a default here if the serializer provides one.

File: pulp_ansible/app/viewsets.py [hunk 27]
```diff
@@ -200,30 +198,33 @@ class AnsibleRepositoryViewSet(RepositoryViewSet, ModifyRepositoryActionMixin):
...
+        sync_kwargs = {"remote_pk": remote.pk, "repository_pk": repository.pk, "mirror": mirror}
+
         if isinstance(remote, RoleRemote):
             sync_func = role_sync
         elif isinstance(remote, CollectionRemote):
             sync_func = collection_sync
+            sync_kwargs["optimize"] = serializer.validated_data.get("optimize", False)
```
Comment (bmbouter) [inline pulp_ansible/app/viewsets.py:222]: Similar here, I don't think this should have a default.

File: pulp_ansible/app/tasks/collections.py (hunk 24 above)
Comment (fao89) [reply pulp_ansible/app/tasks/collections.py:689]: @daviddavis wdyt?

File: pulp_ansible/app/viewsets.py (hunk 21 above)
Comment (bmbouter) [reply pulp_ansible/app/viewsets.py:253]: This looks great

File: pulp_ansible/app/tasks/collections.py [hunk 28]
```diff
@@ -608,13 +663,55 @@ async def _find_all_collections(self):
...
+        if api_version == 3:
+            downloader = self.remote.get_downloader(
+                url=root, silence_errors_for_response_status_codes={404}
+            )
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                return False
+
+            self.last_synced_metadata_time = parse_datetime(metadata.get("published", ""))
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:678]: In cases where there is no published I think the op-op must return False. So we can be more explicit here with:
```
try:
    self.last_synced_metadata_time = parse_datetime(metadata["published"])
except KeyError:
    return False
```

File: pulp_ansible/app/tasks/collections.py [hunk 29]
```diff
@@ -608,13 +663,55 @@ async def _find_all_collections(self):
 
         await asyncio.gather(*tasks)
 
+    async def _can_be_optimized(self):
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:666]: I think this name is confusing. I think can be optimized is the user desire to optimize or not. This is really can we skip the sync or not. With that I'm suggesting:

```suggestion
    async def _should_we_sync(self):
```
Then invert the false and true statements to align with the logic called out by the name.

File: pulp_ansible/app/tasks/collections.py (hunk 24 above)
Comment (daviddavis) [reply pulp_ansible/app/tasks/collections.py:689]: > So that would have it be reporting in all cases unless optimize=False at which point I expect Pulp not to evaluate the no-op case at all.

This makes sense to me.

> I'm suggesting to make this a report that the no-op was checked, not that it occurred.

Would this also contain the result of the no-op check? If not, is there some other way a user could determine if a sync task was no-op?

Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py:689]: With what I imagined (which the test would actually be asserting on) the only way the user would know is they would see the progress report that the no-op was checked and that there are no other progress reports, e.g. metadata downloading, etc. That's not really explicit though.

Comment (daviddavis) [reply pulp_ansible/app/tasks/collections.py:689]: It's not explicit but it's probably still obvious enough for users. I don't have a strong opinion here other than thinking we should show something to users even if it's just showing that there was a no-op check performed.

File: pulp_ansible/app/tasks/collections.py [hunk 30]
```diff
@@ -64,7 +65,43 @@
 log = logging.getLogger(__name__)
 
 
-def sync(remote_pk, repository_pk, mirror):
+def update_collection_remote(remote_pk, *args, **kwargs):
+    """
+    Update a CollectionRemote.
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:70]: Can this docstring get small paragraph after that describes the other behaviors it does besides updating a CollectionRemote?

File: pulp_ansible/app/tasks/collections.py [hunk 31]
```diff
@@ -608,13 +663,60 @@ async def _find_all_collections(self):
 
         await asyncio.gather(*tasks)
 
+    async def _should_we_sync(self):
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:666]: I think the True/False return statements in this function need to be inverted.

File: pulp_ansible/app/tasks/collections.py [hunk 32]
```diff
@@ -608,13 +663,60 @@ async def _find_all_collections(self):
...
+                return True
+
+        return False
+
     async def run(self):
         """
         Build and emit `DeclarativeContent` from the ansible metadata.
         """
+        if self.repository.remote and self.remote.pk == self.repository.remote.pk:
+            can_be_optimized = await self._should_we_sync()
```
Comment (bmbouter) [inline pulp_ansible/app/tasks/collections.py:700]:
```suggestion
            should_we_sync = await self._should_we_sync()
```
This name aligns with the True/False value it's receiving.

**Comment by fao89** [comment]
PR already got unreadable with all the comments,
closing in favor of https://github.com/pulp/pulp_ansible/pull/521
